File: 01/Part1.lua
function sumMatch(str)
  local list, sum = {}, 0
  for char in str:gmatch('%d') do
    list[#list+1] = tonumber(char)
  end
  for i, digit in ipairs(list) do
    local next = list[i+1]
    if next then
      sum = digit == next and (sum+digit) or (sum)
    else
      sum = digit == list[1] and (sum+digit) or (sum)
    end
  end

  return sum
end

local file = io.open('list.txt')
print(sumMatch(file:read('*all'):gsub('[^%d]','')))
file:close()

File: 01/Part1.ts
import * as fs from 'fs'

function sumMatch(str: string) {
  let list = str.split('').map( s => Number.parseInt(s))
  let sum = 0
  for (let index in list) {
    let place = Number.parseInt(index)
    let digit = list[place]
    let other = place === list.length-1 ? list[0] : list[place+1]
    sum += digit === other ? digit : 0
  }
  return sum
}

console.log(sumMatch(fs.readFileSync('list.txt','utf8').replace(/[^0-9]/g,'')))